### DISCUSSION TITLE
Can someone help me understand runtime exception of `ClassNotFoundException` for `org.slf4j.event.LoggingEvent`?

### DISCUSSION THREAD

**Discussion by ayaohsu** [logfellow/logstash-logback-encoder#1001] [General]
Hi, I am using `logstash-logback-encode` 6.6, and my application throws the following runtime exception:

```
Caused by: java.lang.ClassNotFoundException: org.slf4j.event.LoggingEvent
	at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
```

I know it has to do with maven dependencies, and the exception goes away after doing the following upgrade:
"org.slf4j" & "slf4j-api" & "1.7.7" -> "org.slf4j" & "slf4j-api" & "1.7.36"

But I would like to know why this is a runtime exception, instead of just failing at the compile time?

From my understanding, `ClassNotFoundException` is thrown when the classloader cannot find it from the classpath. 
However, wouldn't this be something that a compiler checks, during some sort of linking process?

I am pretty new to Java and would appreciate any guidance / insights provided. 

Thank you for your help!

**Comment by philsttr** [comment]
> But I would like to know why this is a runtime exception, instead of just failing at the compile time?

When compiling your application's source code, the compiler only checks things _directly_ referenced by your application's source code.  Your application's source code most likely does not _directly_ reference `org.slf4j.event.LoggingEvent`, so the compiler does not check for its existence.

`org.slf4j.event.LoggingEvent` was added in slf4j 1.7.15

Your application is using logstash-logback-encoder 6.6, [which was compiled against logback-core 1.2.3](https://github.com/logfellow/logstash-logback-encoder/blob/logstash-logback-encoder-6.6/pom.xml#L25C33-L25C38), [which was compiled against slf4j 1.7.25](https://github.com/qos-ch/logback/blob/v_1.2.3/pom.xml#L60), which defines `org.slf4j.event.LoggingEvent`.  Since logback-core directly uses `org.slf4j.event.LoggingEvent`, the compiler checked for its existence when compiling logback-core.

Later, when the compiler compiles your application's source code against logstash-logback-encoder 6.6, logback-core 1.2.3, and slf4j 1.7.7, the compiler did not report any problems, because your application's source code did not reference `org.slf4j.event.LoggingEvent`.

However, at runtime, when your application runs with logstash-logback-encoder 6.6, logback-core 1.2.3, and slf4j 1.7.7 on the classpath, the jvm will throw the ClassNotFoundException because it loaded the bytecode for a class from logback-core 1.2.3 which requires `org.slf4j.event.LoggingEvent` from slf4j.

**Reply by ayaohsu** [reply]
Thank you so much for the clear explanation!
You are right about my application code not referring to `org.slf4j.event.LoggingEvent` directly so the compiler didn't check for that.
